File: src/main/sponsor/SponsorService.ts
import { net } from 'electron'
import { SponsorBatchSchema, type SponsorStatus, type SponsoredTile, PLACEMENTS } from '@shared/types/sponsor'
import { acceptCreative, reportUrlFor, selectLive } from './sponsorRules'
import type { Database } from '../db/Database'
import type { SettingsStore } from '../settings/SettingsStore'
import { createLogger } from '../logger'

const log = createLogger('sponsor')

/** Refetch no more often than this, however many windows are opened. */
const MIN_FETCH_INTERVAL_MS = 6 * 60 * 60 * 1000
/** A batch is dropped after this even if the server claimed longer. */
const MAX_BATCH_AGE_MS = 7 * 24 * 60 * 60 * 1000

interface TileRow {
  id: string
  sponsor: string
  headline: string
  body: string
  image: string
  click_url: string
  placement: string
  starts_at: number | null
  ends_at: number | null
}

/**
 * Sponsored tiles on the start page.
 *
 * This is the one place Slash shows advertising, and the shape of it is chosen
 * so that it cannot become the thing this browser blocks:
 *
 *  - **Batched, never per-impression.** Creatives are fetched ahead of time and
 *    chosen on-device. The fetch carries no identifier, no browsing data and no
 *    targeting parameters — the sponsor learns that a copy of Slash asked for
 *    tiles, and nothing about who is running it. Showing a tile makes no
 *    request at all.
 *  - **Images are inlined.** A remote `<img src>` would be a per-impression
 *    call to the sponsor's server — a tracking pixel wearing a different hat —
 *    and would undo the batching entirely. Anything that is not a `data:` URL
 *    is dropped at fetch time rather than trusted.
 *  - **Aggregate reporting only.** Impressions and clicks are counted per
 *    creative per *day*. There is deliberately no timestamp finer than that, no
 *    page, no session and no id, because that is the difference between "this
 *    creative was seen 40 times" and a record of when someone opened a tab.
 *  - **Off unless asked for, twice.** It needs both the user's switch and an
 *    operator-configured endpoint. With either missing, no request is ever made.
 */
export class SponsorService {
  private lastFetch = 0
  private rotation = 0

  constructor(
    private readonly db: Database,
    private readonly settings: SettingsStore,
    /**
     * Announces that the cached batch changed.
     *
     * Optional so every probe and test can build this without a window to
     * broadcast to. Without it the start page never learns a batch arrived —
     * see the note on the `sponsor:changed` contract.
     */
    private readonly onChanged: () => void = () => {}
  ) {}

  private get endpoint(): string {
    return this.settings.getAll().sponsorEndpoint.trim()
  }

  private get active(): boolean {
    return this.settings.getAll().sponsoredTilesEnabled && this.endpoint !== ''
  }

  /**
   * What the start page and the settings panel need to know.
   *
   * **Side-effect free.** It used to rotate the batch as it read it, which made
   * every caller advance the ad — including the click handler, which then
   * compared the clicked id against a *different* creative, failed its own
   * guard, and recorded a billable click that opened nothing. Reading state must
   * not change it; rotation is now `recordImpression`'s job, because an advert
   * having been *shown* is the thing that should move the batch on.
   */
  status(): SponsorStatus {
    const cached = this.cachedTiles()
    return {
      enabled: this.settings.getAll().sponsoredTilesEnabled,
      configured: this.endpoint !== '',
      tile: this.active ? this.currentTile(cached.filter((t) => t.placement === 'tile')) : null,
      // No rotation: the backdrop is sold exclusively, so at most one campaign
      // can be live at a time and picking "the first one due" is the whole of
      // the decision.
      background: this.active ? this.livePlacement(cached, 'background') : null,
      banner: this.active ? this.livePlacement(cached, 'banner') : null,
      notice: this.active ? this.livePlacement(cached, 'notice') : null,
      // Two slots, filled in cache order. `livePlacement` answers with one, so
      // the pair is taken directly — a rail campaign is bought as "a gutter",
      // not as "the left gutter", and which side it lands on is not sold.
      rails: this.active ? this.liveRails(cached) : [],
      cached: cached.length,
      pendingReports: this.pendingCount()
    }
  }

  /**
   * The creative currently due, without advancing.
   *
   * Round-robin rather than random, so a small batch is shown evenly instead of
   * one creative dominating by luck — which is what a sponsor is paying for.
   *
   * Campaigns outside their window are skipped here rather than dropped from
   * the cache, so one fetch can carry days of scheduling — which is what makes
   * selling an advert by the hour work against a six-hourly refresh.
   */
  private currentTile(tiles: SponsoredTile[]): SponsoredTile | null {
    return selectLive(tiles, this.rotation, Date.now())
  }

  /**
   * One cached creative by id.
   *
   * How a click resolves its destination: from our own record, by the id the
   * renderer sent, never from whatever the rotation happens to be pointing at.
   *
   * Deliberately ignores the campaign window. A click a moment after a campaign
   * ends is a click on an advert that was legitimately on screen when it was
   * pressed, and the reader should land where the tile said — not on nothing
   * because a timer turned over between the paint and the press.
   */
  /**
   * The one creative live in a placement now, or null.
   *
   * No rotation: every placement except the tile is sold with a small
   * concurrency cap, so the first one due is the whole of the decision.
   */
  /** Up to two live rail creatives, in cache order. */
  private liveRails(cached: SponsoredTile[]): SponsoredTile[] {
    const now = Date.now()
    return cached
      .filter((tile) => tile.placement === 'rail')
      .filter(
        (tile) =>
          (tile.startsAt === null || tile.startsAt <= now) &&
          (tile.endsAt === null || tile.endsAt > now)
      )
      .slice(0, 2)
  }

  private livePlacement(
    cached: SponsoredTile[],
    placement: SponsoredTile['placement']
  ): SponsoredTile | null {
    return selectLive(
      cached.filter((creative) => creative.placement === placement),
      0,
      Date.now()
    )
  }

  tileFor(id: string): SponsoredTile | null {
    return this.cachedTiles().find((tile) => tile.id === id) ?? null
  }

  private cachedTiles(): SponsoredTile[] {
    const now = Date.now()
    return this.db.connection
      .prepare<[number], TileRow>(
        `SELECT id, sponsor, headline, body, image, click_url, placement, starts_at, ends_at
           FROM sponsored_tiles WHERE expires_at > ? ORDER BY id`
      )
      .all(now)
      .map((row) => ({
        id: row.id,
        sponsor: row.sponsor,
        headline: row.headline,
        body: row.body,
        image: row.image,
        clickUrl: row.click_url,
        placement: asPlacement(row.placement),
        startsAt: row.starts_at,
        endsAt: row.ends_at
      }))
  }

  /**
   * Impressions and clicks are recorded per creative per day, nothing finer.
   *
   * Recording an impression is also what **advances the rotation**: the advert
   * has now been shown, so the next start page should get the next one. Keeping
   * it here rather than in `status()` means reading the state never changes it.
   */
  recordImpression(tileId: string): void {
    if (!this.bump(tileId, 'impressions')) return
    this.rotation += 1
  }

  recordClick(tileId: string): void {
    this.bump(tileId, 'clicks')
  }

  /**
   * @returns whether the count was recorded.
   *
   * Ids are checked against the cache first. The id arrives from the renderer,
   * and an unrecognised one would otherwise be inserted and then reported to
   * the sponsor as billing data for a creative that was never served.
   */
  private bump(tileId: string, column: 'impressions' | 'clicks'): boolean {
    if (!this.active) return false
    if (!this.tileFor(tileId)) {
      log.warn(`ignored a ${column} count for unknown creative ${tileId}`)
      return false
    }
    const day = new Date().toISOString().slice(0, 10)
    this.db.connection
      .prepare(
        `INSERT INTO sponsored_counts (tile_id, day, ${column}) VALUES (?, ?, 1)
         ON CONFLICT(tile_id, day) DO UPDATE SET ${column} = ${column} + 1`
      )
      .run(tileId, day)
    return true
  }

  private pendingCount(): number {
    const row = this.db.connection
      .prepare<[], { n: number }>('SELECT COUNT(*) AS n FROM sponsored_counts')
      .get()
    return row?.n ?? 0
  }

  /**
   * Fetches a fresh batch if one is due.
   *
   * Silent on failure by design: an unreachable sponsor endpoint is not the
   * user's problem and must never interrupt browsing. The previous batch keeps
   * showing until it expires, and then nothing shows.
   */
  async refresh(): Promise<void> {
    if (!this.active) return

    // Reporting runs on its own clock, BEFORE the fetch-due check.
    //
    // It used to sit inside it, which meant counts only left when a new batch
    // was due — every six hours. A browser that is never open that long would
    // have accumulated impressions forever and reported none of them, quietly
    // costing whoever sells the inventory their billing data. Caught by the
    // probe rather than in production, which is the whole reason it exists.
    //
    // Its own try/catch, because it now sits outside the one below: an
    // unreachable endpoint must not reject out of here into the floating call
    // at startup.
    try {
      await this.report()
    } catch (error) {
      log.warn('could not report sponsored-tile counts; keeping them for next time', error)
    }

    if (Date.now() - this.lastFetch < MIN_FETCH_INTERVAL_MS) return
    this.lastFetch = Date.now()

    try {
      const response = await net.fetch(this.endpoint, {
        method: 'GET',
        // No cookies, no cache, no credentials. This request must be
        // indistinguishable between one copy of Slash and the next.
        credentials: 'omit',
        cache: 'no-store',
        headers: { accept: 'application/json' }
      })
      if (!response.ok) {
        log.warn(`sponsor endpoint returned ${response.status}`)
        return
      }

      const parsed = SponsorBatchSchema.safeParse(await response.json())
      if (!parsed.success) {
        log.warn('sponsor batch did not match the documented shape; ignored')
        return
      }

      // An operator cannot extend a batch's life indefinitely by claiming a
      // distant expiry: stale advertising is a support problem, not a feature.
      const expiresAt = Math.min(parsed.data.expiresAt, Date.now() + MAX_BATCH_AGE_MS)

      // Rejection rules live in `sponsorRules` so they can be tested without
      // launching a browser — they are the rules that keep a tile from becoming
      // a tracking pixel, which makes them the ones worth pinning down.
      const accepted = parsed.data.tiles.filter((tile) => {
        const verdict = acceptCreative(tile)
        if (!verdict.ok) log.warn(`tile ${tile.id} dropped: ${verdict.reason}`)
        return verdict.ok
      })

      const write = this.db.connection.transaction(() => {
        this.db.connection.prepare('DELETE FROM sponsored_tiles').run()
        const insert = this.db.connection.prepare(
          `INSERT INTO sponsored_tiles
             (id, sponsor, headline, body, image, click_url, fetched_at, expires_at,
              starts_at, ends_at, placement)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
        )
        for (const tile of accepted) {
          insert.run(
            tile.id,
            tile.sponsor,
            tile.headline,
            tile.body,
            tile.image,
            tile.clickUrl,
            Date.now(),
            expiresAt,
            tile.startsAt,
            tile.endsAt,
            tile.placement
          )
        }
      })
      write()
      log.info(`cached ${accepted.length} sponsored tile(s)`)
      // Tell the start page. It reads status once when it mounts, and it mounts
      // at launch — long before any batch has been fetched — so without this
      // the banner and the background stayed invisible until the next restart.
      this.onChanged()
    } catch (error) {
      log.warn('could not refresh sponsored tiles', error)
    }
  }

  /**
   * Sends the aggregate counts and clears them.
   *
   * Counts are only deleted once the server has accepted them, so a failed
   * report costs a retry rather than a sponsor's billing data.
   */
  private async report(): Promise<void> {
    const rows = this.db.connection
      .prepare<[], { tile_id: string; day: string; impressions: number; clicks: number }>(
        'SELECT tile_id, day, impressions, clicks FROM sponsored_counts'
      )
      .all()
    if (rows.length === 0) return

    const reportUrl = reportUrlFor(this.endpoint)
    if (!reportUrl) return

    const response = await net.fetch(reportUrl, {
      method: 'POST',
      credentials: 'omit',
      cache: 'no-store',
      headers: { 'content-type': 'application/json' },
      // Exactly this and nothing else. No install id, no version, no locale —
      // each of which would narrow "some copy of Slash" toward "this one".
      body: JSON.stringify({
        counts: rows.map((row) => ({
          tileId: row.tile_id,
          day: row.day,
          impressions: row.impressions,
          clicks: row.clicks
        }))
      })
    })

    if (!response.ok) {
      log.warn(`report rejected with ${response.status}; counts kept for next time`)
      return
    }
    this.db.connection.prepare('DELETE FROM sponsored_counts').run()
  }

  /** Everything this feature holds, for the user's "forget it all" button. */
  clear(): void {
    this.db.connection.prepare('DELETE FROM sponsored_tiles').run()
    this.db.connection.prepare('DELETE FROM sponsored_counts').run()
    this.lastFetch = 0
    this.onChanged()
  }
}

/**
 * A stored placement string, narrowed.
 *
 * Anything unrecognised becomes a tile. A row written by a newer build than
 * this one must degrade to the smallest, least intrusive shape rather than
 * to whatever the enum happens to list first.
 */
/**
 * A stored placement, checked against the one list of them.
 *
 * This was the **third** copy of that list — after the two schemas — and it
 * failed the same way they did: a value it had not been taught about silently
 * became `'tile'`, so rail campaigns were written correctly, read back as
 * tiles, and appeared in the wrong placement while their own slot stayed
 * empty. Nothing threw and nothing logged. Driving it from `PLACEMENTS` means
 * adding a placement is one edit, in one file.
 */
function asPlacement(value: string): SponsoredTile['placement'] {
  return (PLACEMENTS as readonly string[]).includes(value)
    ? (value as SponsoredTile['placement'])
    : 'tile'
}

File: mobile/android/app/src/main/java/dev/slash/browser/sponsor/SponsorService.kt
package dev.slash.browser.sponsor

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import dev.slash.browser.settings.SettingsStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL

data class SponsoredTile(
    val id: String,
    val sponsor: String,
    val headline: String,
    val body: String,
    /** A `data:` URL. Never a remote address — see `accept` below. */
    val image: String,
    val clickUrl: String,
    val placement: String,
    val startsAt: Long,
    val endsAt: Long
) {
    /**
     * The shape `sponsorRules.acceptCreative` expects.
     *
     * One place that knows the wire format is snake_case and the shared rules
     * are camelCase, so the translation cannot be forgotten at a second call
     * site the way it was at the first.
     */
    fun asCreative(): JSONObject = JSONObject()
        .put("id", id)
        .put("image", image)
        .put("clickUrl", clickUrl)
        .put("startsAt", if (startsAt > 0) startsAt else JSONObject.NULL)
        .put("endsAt", if (endsAt in 1 until Long.MAX_VALUE) endsAt else JSONObject.NULL)

    /**
     * Decoded once and kept, because a start page that re-decodes a base64
     * image on every recomposition is doing real work to draw the same pixels.
     */
    val bitmap: Bitmap? by lazy {
        runCatching {
            val comma = image.indexOf(',')
            if (comma < 0) return@runCatching null
            val bytes = Base64.decode(image.substring(comma + 1), Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        }.getOrNull()
    }
}

/**
 * Sponsored tiles on the start page.
 *
 * The privacy shape is the desktop's, and it is the whole design:
 *
 *  - **Batched, never per-impression.** Creatives are fetched ahead of time and
 *    chosen on-device. The fetch carries no identifier, no browsing data and
 *    nothing about which tile was shown.
 *  - **Images are inlined `data:` URLs.** A remote `<img src>` would be a
 *    request to the sponsor's server every time the tile appeared — a tracking
 *    pixel in a different hat, and it would defeat batching entirely.
 *  - **Inert without an endpoint.** With `sponsorEndpoint` empty, no request is
 *    ever made, so a fresh install contacts nothing.
 *
 * `acceptCreative` is not re-implemented here: it comes from the shared core, so
 * the two platforms cannot drift on a rule that is a privacy promise rather than
 * a formatting preference.
 */
class SponsorService(
    private val context: android.content.Context,
    private val settings: SettingsStore,
    private val acceptCreative: suspend (JSONObject) -> Boolean
) {
    var tiles by mutableStateOf<List<SponsoredTile>>(emptyList())
        private set

    var lastError by mutableStateOf<String?>(null)
        private set

    private var lastFetchAt = 0L
    private var rotation = 0

    /**
     * Counts, aggregated per tile per day, in memory and on disk.
     *
     * **Aggregated, never per-event.** A request at the moment of each
     * impression would tell the sponsor when a specific person looked at their
     * advert, which is precisely the tracking this design exists to avoid. A
     * daily total for a tile says how many times it was shown and nothing about
     * who or when.
     */
    private val counts = linkedMapOf<Pair<String, String>, IntArray>()

    private val cacheFile by lazy { java.io.File(context.filesDir, "sponsors.json") }

    private val endpoint: String get() = settings.current.sponsorEndpoint.trim()

    /**
     * Only the endpoint decides.
     *
     * There is no off switch: sponsored placements are what funds Slash, and a
     * setting that turns off the revenue is not a setting, it is an uninstall
     * with extra steps. What remains configurable is the *endpoint*, so a build
     * with none configured still contacts nothing — which is what keeps a fresh
     * install silent rather than what lets somebody opt out.
     */
    val isActive: Boolean get() = endpoint.isNotEmpty()

    /** The tile to show now, or null. Rotates so one sponsor does not own the slot. */
    fun tileFor(placement: String, now: Long = System.currentTimeMillis()): SponsoredTile? {
        val live = tiles.filter { it.placement == placement && now >= it.startsAt && now < it.endsAt }
        if (live.isEmpty()) return null
        return live[(rotation % live.size + live.size) % live.size]
    }

    fun rotate() {
        rotation += 1
    }

    /**
     * Records that a tile was shown.
     *
     * Idempotent per composition would be wrong — a tile genuinely shown twice
     * is two impressions — but this is called from the start page, which
     * recomposes for unrelated reasons, so the caller passes a key that changes
     * only when the tile actually appears.
     */
    fun noteImpression(tileId: String, now: Long = System.currentTimeMillis()) {
        bump(tileId, now, impressions = 1, clicks = 0)
    }

    fun noteClick(tileId: String, now: Long = System.currentTimeMillis()) {
        bump(tileId, now, impressions = 0, clicks = 1)
    }

    private fun bump(tileId: String, now: Long, impressions: Int, clicks: Int) {
        val day = dayOf(now)
        val row = counts.getOrPut(tileId to day) { intArrayOf(0, 0) }
        row[0] += impressions
        row[1] += clicks
    }

    private fun dayOf(millis: Long): String =
        java.text.SimpleDateFormat("yyyy-MM-dd", java.util.Locale.US)
            .apply { timeZone = java.util.TimeZone.getTimeZone("UTC") }
            .format(java.util.Date(millis))

    /**
     * Sends the aggregated counts.
     *
     * Kept on failure rather than dropped: an advertiser being under-reported
     * because a phone was on a train is somebody's money. Cleared only once the
     * server has accepted them.
     */
    suspend fun reportCounts() = withContext(Dispatchers.IO) {
        if (counts.isEmpty()) return@withContext
        val base = endpoint
        if (base.isEmpty()) return@withContext
        val reportUrl = runCatching { URL(URL(base), "report").toString() }.getOrNull()
            ?: return@withContext

        val payload = JSONObject().put(
            "counts",
            JSONArray().apply {
                for ((key, row) in counts) {
                    put(
                        JSONObject()
                            .put("tileId", key.first)
                            .put("day", key.second)
                            .put("impressions", row[0])
                            .put("clicks", row[1])
                    )
                }
            }
        )

        try {
            val connection = (URL(reportUrl).openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                useCaches = false
                connectTimeout = 15_000
                readTimeout = 20_000
                setRequestProperty("content-type", "application/json")
            }
            try {
                connection.outputStream.use { it.write(payload.toString().toByteArray()) }
                val code = connection.responseCode
                if (code !in 200..299) {
                    Log.w(TAG, "report rejected with $code; counts kept for next time")
                    return@withContext
                }
            } finally {
                connection.disconnect()
            }
            counts.clear()
        } catch (error: Exception) {
            Log.w(TAG, "could not report sponsor counts", error)
        }
    }

    /**
     * Fetches a batch, at most every [MIN_INTERVAL_MS].
     *
     * Silent on failure by design: an unreachable sponsor endpoint is not the
     * user's problem and must not become a visible error on the page they open
     * most. The start page simply shows nothing where a tile would be.
     */
    suspend fun refresh(force: Boolean = false) = withContext(Dispatchers.IO) {
        if (!isActive) {
            tiles = emptyList()
            return@withContext
        }
        val now = System.currentTimeMillis()
        if (!force && now - lastFetchAt < MIN_INTERVAL_MS) return@withContext
        lastFetchAt = now

        try {
            val body = get(endpoint)
            val array = body.optJSONArray("creatives")
                ?: body.optJSONArray("items")
                ?: JSONArray()

            val accepted = mutableListOf<SponsoredTile>()
            for (i in 0 until array.length()) {
                val row = array.optJSONObject(i) ?: continue
                val tile = parse(row) ?: continue
                // Asked of the shared core rather than decided here — but in
                // the shape the shared rules expect. `Creative` is camelCase
                // (`clickUrl`, `startsAt`); the wire format is snake_case, and
                // passing the raw row meant `clickUrl` was undefined, so the
                // https test refused *every* creative including the valid ones.
                if (!acceptCreative(tile.asCreative())) {
                    Log.w(TAG, "creative ${tile.id} refused by the shared rules")
                    continue
                }
                accepted += tile
            }
            tiles = accepted
            writeCache(accepted)
            lastError = null
            Log.i(TAG, "sponsor batch: ${accepted.size} accepted of ${array.length()}")
        } catch (error: Exception) {
            lastError = error.message
            Log.w(TAG, "sponsor fetch failed", error)
        }
    }

    /**
     * Reads the last batch from disk.
     *
     * A batch is bought by the hour and fetched at most every six, so holding it
     * only in memory meant every cold start either showed nothing or made a
     * fetch — the first cheats the advertiser out of hours they paid for, the
     * second turns "batched every six hours" into "on every launch".
     *
     * Creatives are **re-vetted on load**, not trusted because they were vetted
     * once: the file is on disk, and a rule that only runs at fetch time is a
     * rule that can be walked around by editing it.
     */
    suspend fun loadCache() = withContext(Dispatchers.IO) {
        if (!isActive) return@withContext
        val text = runCatching { cacheFile.takeIf { it.exists() }?.readText() }.getOrNull()
            ?: return@withContext
        runCatching {
            val stored = JSONObject(text)
            lastFetchAt = stored.optLong("fetchedAt")
            val array = stored.optJSONArray("creatives") ?: JSONArray()
            val restored = mutableListOf<SponsoredTile>()
            for (i in 0 until array.length()) {
                val row = array.optJSONObject(i) ?: continue
                val tile = parse(row) ?: continue
                if (!acceptCreative(tile.asCreative())) continue
                restored += tile
            }
            tiles = restored
        }.onFailure { Log.w(TAG, "could not read the sponsor cache", it) }
    }

    private fun writeCache(accepted: List<SponsoredTile>) {
        runCatching {
            val array = JSONArray()
            for (tile in accepted) {
                array.put(
                    JSONObject()
                        .put("id", tile.id)
                        .put("sponsor", tile.sponsor)
                        .put("headline", tile.headline)
                        .put("body", tile.body)
                        .put("image", tile.image)
                        .put("click_url", tile.clickUrl)
                        .put("placement", tile.placement)
                        .put("starts_at", tile.startsAt)
                        .put("ends_at", tile.endsAt)
                )
            }
            cacheFile.writeText(
                JSONObject().put("fetchedAt", lastFetchAt).put("creatives", array).toString()
            )
        }.onFailure { Log.w(TAG, "could not write the sponsor cache", it) }
    }

    /** Everything this feature holds, for a "forget it all" action. */
    fun clear() {
        tiles = emptyList()
        counts.clear()
        lastFetchAt = 0
        runCatching { cacheFile.delete() }
    }

    private fun parse(row: JSONObject): SponsoredTile? {
        val id = row.optString("id").ifEmpty { return null }
        return SponsoredTile(
            id = id,
            sponsor = row.optString("sponsor"),
            headline = row.optString("headline"),
            body = row.optString("body"),
            image = row.optString("image"),
            clickUrl = row.optString("click_url").ifEmpty { row.optString("clickUrl") },
            placement = row.optString("placement").ifEmpty { "tile" },
            startsAt = row.optLong("starts_at", row.optLong("startsAt")),
            endsAt = row.optLong("ends_at", row.optLong("endsAt", Long.MAX_VALUE))
        )
    }

    private fun get(url: String): JSONObject {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 20_000
        }
        return try {
            val code = connection.responseCode
            val stream = if (code in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader()?.use(BufferedReader::readText).orEmpty()
            if (code !in 200..299) throw IllegalStateException("sponsor endpoint returned $code")
            if (text.isBlank()) JSONObject() else JSONObject(text)
        } finally {
            connection.disconnect()
        }
    }

    private companion object {
        const val TAG = "SlashSponsor"

        /** Refetch no more often than this, however many times the page is opened. */
        const val MIN_INTERVAL_MS = 6 * 60 * 60 * 1000L
    }
}
